File: SS/Labs/Lab3/RDP.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

char input[100];
int pos = 0;
int error = 0;

void E();
void E_prime();
void T();
void T_prime();
void F();

void skip_space() {
    while (input[pos] == ' ' || input[pos] == '\t')
        pos++;
}

void match(char c) {
    skip_space();
    if (input[pos] == c)
        pos++;
    else
        error = 1;
}

// E -> T E'
void E() {
    if (error) return;
    T();
    E_prime();
}

// E' -> + T E' | eps
void E_prime() {
    if (error) return;
    skip_space();
    if (input[pos] == '+') {
        pos++;
        T();
        E_prime();
    }
    // else epsilon, do nothing
}

// T -> F T'
void T() {
    if (error) return;
    F();
    T_prime();
}

// T' -> * F T' | epsilon
void T_prime() {
    if (error) return;
    skip_space();
    if (input[pos] == '*') {
        pos++;
        F();
        T_prime();
    }
    // else epsilon, do nothing
}

// F -> ( E ) | id
void F() {
    if (error) return;
    skip_space();
    
    // Check for ( E )
    if (input[pos] == '(') {
        pos++;
        E();
        match(')');
    }
    // Check for id (identifier - letter followed by alphanumeric)
    else if (isalpha(input[pos])) {
        while (isalnum(input[pos]))
            pos++;
    }
    // Check for number
    else if (isdigit(input[pos])) {
        while (isdigit(input[pos]))
            pos++;
    }
    else {
        error = 1;
    }
}

int main() {
    printf("Enter expression: ");
    fgets(input, sizeof(input), stdin);
    input[strlen(input)-1] = '\0';  // remove newline
    
    printf("\nGrammar (after left recursion elimination):\n");
    printf("E  -> T E'\n");
    printf("E' -> + T E' | epsilon\n");
    printf("T  -> F T'\n");
    printf("T' -> * F T' | epsilon\n");
    printf("F  -> ( E ) | id\n\n");
    E();
    skip_space();
    printf("Result: ");
    if (!error && input[pos] == '\0')
        printf("VALID expression (Accepted)\n");
    else
        printf("INVALID expression (Rejected)\n");
    return 0;
}

File: SS/Labs/PythonLabs/Lab3/RDP.py
# Lab 3 — Recursive Descent Parser (RDP)
# Grammar (after left-recursion elimination):
#   E  -> T E'
#   E' -> + T E' | epsilon
#   T  -> F T'
#   T' -> * F T' | epsilon
#   F  -> ( E ) | id | number

class RDP:
    def __init__(self, text):
        self.inp = text.replace(' ', '').replace('\t', '')
        self.pos = 0
        self.error = False

    def peek(self):
        return self.inp[self.pos] if self.pos < len(self.inp) else '\0'

    def E(self):
        if self.error: return
        self.T()
        self.E_prime()

    def E_prime(self):
        if self.error: return
        if self.peek() == '+':
            self.pos += 1
            self.T()
            self.E_prime()

    def T(self):
        if self.error: return
        self.F()
        self.T_prime()

    def T_prime(self):
        if self.error: return
        if self.peek() == '*':
            self.pos += 1
            self.F()
            self.T_prime()

    def F(self):
        if self.error: return
        ch = self.peek()
        if ch == '(':
            self.pos += 1
            self.E()
            if self.peek() == ')':
                self.pos += 1
            else:
                self.error = True
        elif ch.isalpha():
            while self.pos < len(self.inp) and self.inp[self.pos].isalnum():
                self.pos += 1
        elif ch.isdigit():
            while self.pos < len(self.inp) and self.inp[self.pos].isdigit():
                self.pos += 1
        else:
            self.error = True

    def parse(self):
        self.E()
        return not self.error and self.pos == len(self.inp)

def main():
    print("Grammar (after left recursion elimination):")
    print("E  -> T E'")
    print("E' -> + T E' | epsilon")
    print("T  -> F T'")
    print("T' -> * F T' | epsilon")
    print("F  -> ( E ) | id\n")

    expr = input("Enter expression: ").strip()
    parser = RDP(expr)
    result = parser.parse()
    print("Result:", "VALID expression (Accepted)" if result else "INVALID expression (Rejected)")

if __name__ == "__main__":
    main()
